src/herder/TxQueueLimiter.cpp

```diff
@@ -10,244 +10,216 @@
-                         TransactionFrameBasePtr const& oldTx) const
-{
-    // enforce min fee if needed
-    if (mMinFeeNeeded.second != 0)
+                         TransactionFrameBasePtr const& oldTx)
+{
+    // The usage pattern of the limiter is either canAddTx->evictTransactions
+    // (where we cleanup the cached txs to evict) or canAddTx->canAddTx
+    // (when tx hasn't been added for unrelated reasons). Clear the cache here
+    // to address the second pattern.
+    mEvictionCache.clear();
+    // We cannot normally initialize transaction queue in the constructor
+    // because `maxQueueSizeOps()` may not be initialized. Hence we initialize
+    // lazily during the add/reset.
+    if (mTxs == nullptr)
     {
-        auto cmp3Min =
-            feeRate3WayCompare(newTx->getFeeBid(), newTx->getNumOperations(),
-                               mMinFeeNeeded.first, mMinFeeNeeded.second);
-        if (cmp3Min <= 0)
-        {
-            auto minFee = computeBetterFee(newTx, mMinFeeNeeded.first,
-                                           mMinFeeNeeded.second);
-            return std::make_pair(false, minFee);
-        }
-    }
-
-    auto newOps = mQueueSizeOps;
-    if (oldTx)
-    {
-        // oldTx is currently tracked by the queue,
-        // so this should always hold
-        releaseAssert(oldTx->getNumOperations() <= newOps);
-        newOps -= oldTx->getNumOperations();
-    }
-    newOps += newTx->getNumOperations();
-
-    // if there is enough space, return
-    if (newOps <= maxQueueSizeOps())
-    {
-        return std::make_pair(true, 0ll);
-    }
-
-    // need to see if we could be added by kicking out cheaper transactions
-    // starting with the cheapest one
-    auto neededOps = newOps - maxQueueSizeOps();
-    auto id = newTx->getSourceID();
-    for (auto it = mTxs->begin(); it != mTxs->end(); ++it)
-    {
-        if (feeRate3WayCompare(*(*it), *newTx) >= 0)
-        {
-            auto minFee = computeBetterFee(newTx, (*it)->getFeeBid(),
-                                           (*it)->getNumOperations());
-            return std::make_pair(false, minFee);
-        }
-        // ensure that this transaction is not from the same account
-        auto& tx = *it;
-        if (tx->getSourceID() == id)
-        {
-            return std::make_pair(false, 0ll);
-        }
-        auto curOps = tx->getNumOperations();
-        if (neededOps <= curOps)
-        {
-            return std::make_pair(true, 0ll);
-        }
-        neededOps -= curOps;
+        resetTxs();
     }
 
-    // we reach this point if the queue doesn't have capacity for that
-    // transaction even when empty. Combination of multiplier and max ledger
-    // size is too small for whatever reason
-    static size_t lastMax = std::numeric_limits<size_t>::max();
-    if (lastMax != maxQueueSizeOps())
+    // If some transactions were evicted from this or generic lane, make sure
```

Comment:
too many things going on here: you renamed fields, added new ones, and refactored the code all in one go, so I can't really tell which invariant is supposed to hold